### ISSUE TITLE
Remove all docstrings when python is run in optimization mode

### ISSUE THREAD

**Issue by vitrioil** [pytorch/pytorch#76659]
## Issue description
Follow up of #76034.

There are docstrings still present in *pyc files when python is run in optimization mode (`-OO`). 

## Code example

```python
$ /usr/local/bin/python3.9 -OO
Python 3.9.2 (default, Mar 26 2021, 21:58:27) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> help(torch.nn.Conv1d)
```

This has to be separately handled.

Current *pyc file size for pytorch:

1. Without optimization
```bash
$ find . -type f -name '*39.pyc' -exec du -ch {} + | grep total$ 
12M	total
```
2. With optimization (`-O`)
```bash
$ find . -type f -name '*39.opt-1.pyc' -exec du -ch {} + | grep total$
5.3M	total
```
3. With optimization (`-OO`)
```bash
$ find . -type f -name '*39.opt-2.pyc' -exec du -ch {} + | grep total$
3.9M	total
```

I don't see any noticeable time difference while importing torch.

**Comment by albanD** [comment]
Note that the c++ function `THPModule_addDocStr` that is used extensively is also most likely a very big offender here. We use it extensively to add docstring to our c++ bindings and I don't think there is a skip there for when the `__debug__` flag is not set.

**Comment by vitrioil** [comment]
I think we can use this: https://docs.python.org/3/c-api/init.html#c.Py_OptimizeFlag

I see `THPModule_addDocStr` referencing numpy's `arr_add_docstring`.

Numpy is doing something [similar](
https://github.com/numpy/numpy/blob/5cc7ef066fca7a821a2160b095578384c301ae3c/numpy/core/src/multiarray/compiled_base.c#L1404-L1406):
```cpp
    /* Don't add docstrings */
    if (Py_OptimizeFlag > 1) {
        Py_RETURN_NONE;
    }
```

I will check this. I am not sure which method uses this API though.

**Comment by albanD** [comment]
Good catch!
Could you try adding that skip and see what remains in the pyc?

**Comment by vitrioil** [comment]
Oh I just saw this is already handled in pytorch!
Also there is no difference in pyc.
https://github.com/pytorch/pytorch/blob/54c75e1e8f179d101cc4c19b715941159b6b8009/torch/csrc/Module.cpp#L225-L227

**Comment by albanD** [comment]
That code is only getting the two arguments from the function into these two objects.
If this fails it is because you passed the wrong number of arguments and that would raise an error.

**Comment by vitrioil** [comment]
Right, I can see that adding that check removes documentation from methods like:
`from torch._C import _has_torch_function` and many others.

I checked the *pyc size before and after the check and I did not see any difference.
I used `sys.pycache_prefix` to manually change *pyc to a fresh location in both the cases.

It stays at 3928 bytes, is that because the strings are dynamically added? There are no `pyc` files for `pyi` linked files so that is why there is no difference.

**Comment by vitrioil** [comment]
Not sure how many % of docstring I was able to capture, but overall length of docstring (rough estimates):

`Without -OO: 8,848,771`
`With -OO: 2,907,888`

Same for numpy:
`Without -OO: 9,691,105`
`With -OO: 3,545,369`

Source: https://gist.github.com/vitrioil/2831ab71beccd4e5bf6a85754828a6a0
